st: wba65: add stm32wba65i_dk1 board, 'ns' variant for dk1 and nucleo boards

etienne-lms opened the pull request:
This P-R mainly adds 'ns' variant for 2 STM32WBA65xx based ST boards and also adds wkup_pin ST sample on nucleo_wba65ri.

- Add stm32wba65i_dk1 boards and an 'ns' variant for this board.
- Refine nucleo_wba65ri documentation, add wake-from-poweroff demo for this board and the board 'ns' variant
- ~~Add PSA Arch test for WBA65xx based boards ([deps on psa-arch-tests#13](https://github.com/zephyrproject-rtos/psa-arch-tests/pull/13))~~
**Edited: this update is postponed to a later dedicated P-R.**

This P-R is based on the recently merged TF-M/STM32 updates by https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/145 and #93520.

This P-R supersedes https://github.com/zephyrproject-rtos/zephyr/pull/90162.

etienne-lms committed:
boards: st: Add stm32wba65i-dk1

Add stm32wba65i-dk1 board support with UART console, LEDs, joystick keys
using ADC channel 6.

Board YAML file does not list 'supported' tags since the board is
very similar to nucleo_wba65ri for which supported features are already
covered.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
modules: trusted-firmware-m: Add STM32_FLASH_LAYOUT_BEGIN_OFFSET

Add TF-M directive STM32_FLASH_LAYOUT_BEGIN_OFFSET needed to specify
the gap needed by external boot stage resources at flash beginning.
The offset tells STM32 TF-M firmware the base offset in the flash
where the several TF-M and non-secure image areas shall be located.

The CMake directive was introduced mainline TF-M commit [1] and merged
in Zephyr TF-M repository [2].

Link: https://github.com/TrustedFirmware-M/trusted-firmware-m/commit/fc035b874e0ab86e2a13a328da3dce0cf18eb566 [1]
Link: https://github.com/zephyrproject-rtos/trusted-firmware-m/commit/954dc805411be8fedac4ed7ddfefe966ffb35576 [2]
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
boards: st: stm32wba65i_dk1: Add 'ns' variant for TF-M support

Add variant ns to stm32wba65i_dk1 board to embed TF-M in the SoC
secure world. The flash layout is synced with the layout defined
in Zephyr TF-M integration of platform STM32WBA65I.

Successfully tested against a few samples and test samples:
- samples/tfm_integration/psa_crypto
- samples/tfm_integration/psa_protected_storage
- samples/tfm_integration/tfm_ipc
- samples/tfm_integration/tfm_regression_test
- samples/tfm_integration/tfm_secure_partition
- tests/subsys/secure_storage/psa/crypto
- tests/subsys/secure_storage/psa/its
  (with CONFIG_TFM_ITS_MAX_ASSET_SIZE_OVERRIDE=y
  and CONFIG_TFM_ITS_MAX_ASSET_SIZE=256)

Support for PSA Arch Tests (samples/tfm_integration/tfm_psa_test) is
not yet merged but is in under review [1].

Link: https://github.com/ARM-software/psa-arch-tests/pull/406 [1]
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
boards: st: nucleo_wba65ri: Add 'ns' variant for TF-M support

Add variant 'ns' to nucleo_wba65ri board to embed TF-M in the SoC
secure world.

The flash layout is synced with the layout defined in Zephyr TF-M
integration of platform STM32WBA65I.

- samples/tfm_integration/psa_crypto
- samples/tfm_integration/psa_protected_storage
- samples/tfm_integration/tfm_ipc
- samples/tfm_integration/tfm_regression_test
- samples/tfm_integration/tfm_secure_partition
- tests/subsys/secure_storage/psa/crypto
- tests/subsys/secure_storage/psa/its
  (with CONFIG_TFM_ITS_MAX_ASSET_SIZE_OVERRIDE=y
  and CONFIG_TFM_ITS_MAX_ASSET_SIZE=256)

Support for PSA Arch Tests (samples/tfm_integration/tfm_psa_test) is
not yet merged but is in under review [1].

Link: https://github.com/ARM-software/psa-arch-tests/pull/406 [1]
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
modules: trusted-firmware-m: Declare stm32wba65i support

Declare stm32wba65i-dk1 and nucleo_wba65ri boards support in TF-M.
Both comply with TF-M integration of platform stm/stm32wba65i-dk.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
boards: st: nucleo_wba65ri: rename documentation file

Rename nucleo_wba65ri board documentation file for consistency
within Zephyr file tree.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
boards: st: nucleo_wba65ri: fix special characters in board doc

Replace (R) and (TM) special charaters with |reg| and |trade|
aliases in nucleo_wba65ri board documentation file. By the way
also fix some characters case, add info and move Cortex-M description
in the overview section to prevent information duplication.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
boards: st: nucleo_wba65ri: add low power numbers in board doc

Add missing low power consumption numbers in nucleo_wba65ri board
documentation. These numbers are dumped from the SoC datashhet.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
samples: boards: st: power_mgmt: wkup_pins: add WBA65 boards

Add WBA65RI Nucleo board to ST wkup_pins board sample.
By the way, sort the board names in sample.yaml.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

asm5878 commented on boards/st/stm32wba65i_dk1/stm32wba65i_dk1.dts:
```diff
@@ -0,0 +1,187 @@
+&rtc {
+	status = "okay";
+	clocks = <&rcc STM32_CLOCK(APB7, 21)>,
+			<&rcc STM32_SRC_LSE RTC_SEL(1)>;
```
Indentation

asm5878 reviewed (dismissed):
Just few minor remarks on indentation, otherwise LGTM

etienne-lms commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -515,4 +517,12 @@ config TFM_EXCEPTION_INFO_DUMP
 	  On fatal errors in the secure firmware, capture info about the exception.
 	  Print the info if the SPM log level is sufficient.
 
+config TFM_STM32_FLASH_LAYOUT_BEGIN_OFFSET
+	int "Offset gap at beginning of flash layout"
```
Since this applies on STM32 SoC based images, would `depends on SOC_FAMILY_STM32` make sense to you?

I think I should explicit in above and below descriptions that this config is STM32 specific.

tomi-font replied:
Sounds good to me.

tomi-font commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -281,6 +281,12 @@ if (CONFIG_BUILD_WITH_TFM)
     list(APPEND TFM_CMAKE_ARGS -DETHOS_DRIVER_PATH=${CONFIG_TFM_ETHOS_DRIVER_PATH_LOCAL})
   endif()
 
+  if (NOT CONFIG_TFM_STM32_FLASH_LAYOUT_BEGIN_OFFSET EQUAL 0)
```
Maybe here we could first check that `CONFIG_TFM_STM32_FLASH_LAYOUT_BEGIN_OFFSET` is defined?

tomi-font reviewed (commented):
Some more suggestions but the changes under `modules/trusted-firmware-m` look good to me, will let others review the bulk of the changes and approve the PR.

erwango commented on boards/st/stm32wba65i_dk1/stm32wba65i_dk1_stm32wba65xx_ns.yaml:
```diff
@@ -0,0 +1,10 @@
+toolchain:
+  - zephyr
+  - gnuarmemb
+ram: 512
+flash: 2048
```
To be updated vs ns partition size ?

erwango reviewed (commented):
Open questions otherwise LGTM

etienne-lms commented on boards/st/stm32wba65i_dk1/stm32wba65i_dk1_stm32wba65xx_ns_defconfig:
```diff
@@ -0,0 +1,34 @@
+CONFIG_ROM_START_OFFSET=0x400
+
+# Enable TZ non-secure configuration
+CONFIG_ARM_TRUSTZONE_M=y
+CONFIG_RUNTIME_NMI=y
+CONFIG_TRUSTED_EXECUTION_NONSECURE=y
+CONFIG_BUILD_WITH_TFM=y
+CONFIG_TFM_BL2=n
```
Ok, I think the SoC defconfig could enable `TFM_BL2_NOT_SUPPORTED`.
I see `CONFIG_ARM_TRUSTZONE_M=y` is not needed neither, already selected by *soc/st/stm32/stm32wbax/Kconfig*.

etienne-lms committed:
soc: st: stm32wba: TF-M does not support BL2 for WBA65x

Enable TFM_BL2_NOT_SUPPORTED configuration for STM32WBA65x SoC since
TF-M does not implement the BL2 boot stage for this SoC series.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented:
Rebased to fix merge conflict in nucleo_wba65ri board doc. No other changes.

mathieuchopstm commented on boards/st/nucleo_wba65ri/board.cmake:
```diff
@@ -1,6 +1,31 @@
+  endif()
+
+  # System entry point is TF-M vector, located 1kByte after tfm_fmw_partition in DTS
+  get_target_property(TFM_FWM_NODE_NAME devicetree_target "DT_NODELABEL|slot0_secure_partition")
+  string(REGEX REPLACE ".*@([^@]+)$" "\\1" TFM_FWM_OFFSET "${TFM_FWM_NODE_NAME}")
```
Are you sure there isn't a CMake function to read arbitrary DT properties (i.e., `DT_PROP(DT_NODELABEL(slot0_secure_partition), reg)` here)?

etienne-lms replied:
Indeed there is.

mathieuchopstm commented on boards/st/nucleo_wba65ri/doc/index.rst:
```diff
@@ -165,6 +163,47 @@ To fetch Binary Blobs:
+   $ ./build/tfm/api_ns/regression.sh
+   $ west flash
+
+Please note that, after having programmed the board for a TrustZone enabled system
+(e.g. with ``./build/tfm/api_ns/regression.sh``), the SoC TZEN Option Byte is enabled
+and you will need to operate specific sequence to disable this TZEN Option Byte
+configuration to get your board back in normal state for booting with a TrustZone
+disabled system (e.g. without TF-M support).
```
Maybe indicate where this "specific sequence" is documented?

mathieuchopstm commented on boards/st/stm32wba65i_dk1/Kconfig.defconfig:
```diff
@@ -0,0 +1,25 @@
+
+if BUILD_WITH_TFM
+
+# Not defining LIBC malloc arena has the effect of declaring all available RAM
+# as available for malloc.
+# This currently conflicts with TF-M MPU setting, resulting in a hard fault.
+# Define a specific size to avoid this situation.
```
Would it be possible to document how much (or how to estimate) RAM is consumed by TF-M?

mathieuchopstm replied:
Nit: pack on three lines + `available RAM` -> `unused RAM` (clearer IMO)
```suggestion
# Not defining LIBC malloc arena has the effect of declaring all unused RAM
# as available for malloc. This currently conflicts with TF-M MPU setting,
# resulting in a hard fault. Define a specific size to avoid this situation.
```

mathieuchopstm commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -20,6 +20,8 @@ config TFM_BOARD
 	default "stm/b_u585i_iot02a" if BOARD_B_U585I_IOT02A
 	default "stm/nucleo_l552ze_q" if BOARD_NUCLEO_L552ZE_Q
 	default "stm/stm32l562e_dk" if BOARD_STM32L562E_DK
+	default "stm/stm32wba65i_dk" if BOARD_NUCLEO_WBA65RI
+	default "stm/stm32wba65i_dk" if BOARD_STM32WBA65I_DK1
```
Maybe:
```suggestion
	default "stm/stm32wba65i_dk" if BOARD_NUCLEO_WBA65RI || BOARD_STM32WBA65I_DK1
```

But is sharing the same value for different boards expected?

mathieuchopstm commented on boards/st/stm32wba65i_dk1/Kconfig.defconfig:
Ditto for `boards: st: nucleo_wba65ri: Add 'ns' variant for TF-M support`

mathieuchopstm reviewed (dismissed):
Non-blocker, LGTM otherwise

etienne-lms commented:
Comment addressed.

erwango commented:
@d3zd3z PTAL
